### PR TITLE
Tip selector

### LINKED ISSUE #200: Implement tip selection
- [ ] Handle reference

### PR DESCRIPTION
**Pull request by thibault-martinez**
Fixes #200 with the following changes
- Enable use of a reference
- Check consistency of selected tips
- Check errors
- Renamed `tipselection` to `tip_selector`

# Test Plan:
Not applicable

### CODE REVIEW

File: consensus/tip_selector/tip_selector.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+#include "consensus/cw_rating_calculator/cw_rating_calculator.h"
+#include "consensus/entry_point_selector/entry_point_selector.h"
+#include "consensus/exit_probability_randomizer/exit_probability_randomizer.h"
+#include "consensus/exit_probability_validator/exit_probability_validator.h"
+#include "consensus/snapshot/snapshot.h"
+#include "consensus/tip_selector/tip_selector.h"
+#include "utils/logger_helper.h"
+
+#define TIP_SELECTOR_LOGGER_ID "consensus_tipselection"
```
Comment (th0br0) [inline consensus/tip_selector/tip_selector.c]: should be `consensus_tip_selector`?

File: consensus/tip_selector/tip_selector.c [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  flex_trit_t ep_trits[FLEX_TRIT_SIZE_243];
+  trit_array_t ep = {.trits = ep_trits,
+                     .num_trits = HASH_LENGTH_TRIT,
+                     .num_bytes = FLEX_TRIT_SIZE_243,
+                     .dynamic = 0};
+  cw_calc_result rating_results = {.cw_ratings = NULL, .tx_to_approvers = NULL};
+  bool consistent = false;
+  hash_stack_t tips_stack = NULL;
+
+  rw_lock_handle_rdlock(
```
Comment (th0br0) [inline consensus/tip_selector/tip_selector.c:55]: Once we have a working clone, we should strive to eliminate these locks. They are unnecessary with a slight redesign

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: consensus/tip_selector/tip_selector.c (hunk 1 above)
Comment (thibault-martinez) [reply consensus/tip_selector/tip_selector.c]: Yes

File: consensus/tip_selector/tip_selector.c (hunk 2 above)
Comment (thibault-martinez) [reply consensus/tip_selector/tip_selector.c:55]: Agreed

**Review by th0br0 — APPROVED** [review]

**Pull request merged**